File: 349. Intersection of Two Arrays.cpp
class Solution {
public:
    // two pointers: O(max(mlogm, nlogn))time, O(1)space, 58.27%, ok
    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        vector<int> ret;
        if(nums1.empty() || nums2.empty()) return ret;
        sort(nums1.begin(), nums1.end());
        sort(nums2.begin(), nums2.end());
        for(int i = 0, j = 0; i < nums1.size() && j < nums2.size();){
            if(nums1[i] == nums2[j]){
                ret.push_back(nums1[i]);
                while(i+1 < nums1.size() && nums1[i+1] == nums1[i]) i++;
                while(j+1 < nums2.size() && nums2[j+1] == nums2[j]) j++;
                i++; j++;
            }
            else if(nums1[i] < nums2[j]) i++;
            else j++;
        }
        return ret;
    }

    // hashset/hashmap: O(m + n)time, O(m or n)space, 58.27%, ok
    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        vector<int> ret;
        if(nums1.empty() || nums2.empty()) return ret;
        unordered_set<int> elem(nums1.begin(), nums1.end());
        for(auto n: nums2){
            if(elem.count(n)){
                ret.push_back(n);
                elem.erase(n);
            }
        }
        return ret;
    }
};

File: 349. Intersection of Two Arrays.py
class Solution(object):
    
    # brute force: O(mn) time, O(1) space
    def intersection(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        ret = []
        for i in xrange(len(nums1)):
            if nums1[i] in ret:
                continue
            for j in xrange(len(nums2)):
                if nums1[i] == nums2[j]:
                    ret.append(nums1[i])
                    break
        return ret
    
    # set: O(m+n) time, O(m+n) space
    def intersection(self, nums1, nums2):
        set1, set2 = set(nums1), set(nums2)
        return [num for num in set1 if num in set2]
    
    # pythonic set: O(m+n) time, O(m+n) space
    def intersection(self, nums1, nums2):
        set1, set2 = set(nums1), set(nums2)
        return list(set1.intersection(set2))
    
    # hash table: O(m+n) time, O(min(m, n)) space
    def intersection(self, nums1, nums2):
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1
        cnt = {num:1 for num in nums1}
        ret = []
        for num in nums2:
            if num in cnt and num not in ret:
                ret.append(num)
        return ret
        